packages/assets-controllers/src/TokenDetectionController.ts

```diff
@@ -1,242 +1,265 @@
-    this.getBalancesInSingleCall = getBalancesInSingleCall;
-    this.getNetworkClientById = getNetworkClientById;
-
-    onTokenListStateChange(({ tokenList }) => {
-      const hasTokens = Object.keys(tokenList).length;
-
-      if (hasTokens) {
-        this.detectTokens();
-      }
+    super({
+      name: controllerName,
+      messenger,
+      state: {},
+      metadata: {},
     });
 
-    onPreferencesStateChange(({ selectedAddress, useTokenDetection }) => {
-      const {
-        selectedAddress: currentSelectedAddress,
-        isDetectionEnabledFromPreferences,
-      } = this.config;
-      const isSelectedAddressChanged =
-        selectedAddress !== currentSelectedAddress;
-      const isDetectionChangedFromPreferences =
-        isDetectionEnabledFromPreferences !== useTokenDetection;
-
-      this.configure({
-        isDetectionEnabledFromPreferences: useTokenDetection,
-        selectedAddress,
-      });
+    this.setIntervalLength(interval);
+    this.#chainId = chainId;
+    this.#selectedAddress = selectedAddress;
 
-      if (
-        useTokenDetection &&
-        (isSelectedAddressChanged || isDetectionChangedFromPreferences)
-      ) {
-        this.detectTokens();
-      }
-    });
+    this.#getTokensState = getTokensState;
+    this.#addDetectedTokens = addDetectedTokens;
+    this.#getBalancesInSingleCall = getBalancesInSingleCall;
 
-    onNetworkStateChange(({ providerConfig: { chainId } }) => {
-      const { chainId: currentChainId } = this.config;
-      const isDetectionEnabledForNetwork =
-        isTokenDetectionSupportedForNetwork(chainId);
-      const isChainIdChanged = currentChainId !== chainId;
+    this.messagingSystem.unregisterActionHandler(
+      'TokenDetectionController:getState',
+    );
 
-      this.configure({
-        chainId,
-        isDetectionEnabledForNetwork,
-      });
+    this.messagingSystem.subscribe(
+      'TokenListController:stateChange',
+      ({ tokenList }) => {
+        const hasTokens = Object.keys(tokenList).length;
 
-      if (isDetectionEnabledForNetwork && isChainIdChanged) {
-        this.detectTokens();
-      }
-    });
+        if (hasTokens) {
+          this.detectTokens();
+        }
+      },
+    );
+
+    onPreferencesStateChange(
+      ({ selectedAddress: newSelectedAddress, useTokenDetection }) => {
+        const isSelectedAddressChanged =
+          this.#selectedAddress !== newSelectedAddress;
+        const isDetectionChangedFromPreferences =
+          this.#isDetectionEnabledFromPreferences !== useTokenDetection;
+
+        this.#selectedAddress = newSelectedAddress;
+        this.#isDetectionEnabledFromPreferences = useTokenDetection;
+
+        if (
+          useTokenDetection &&
+          (isSelectedAddressChanged || isDetectionChangedFromPreferences)
+        ) {
+          this.detectTokens();
+        }
+      },
+    );
+
+    this.messagingSystem.subscribe(
+      'NetworkController:stateChange',
```

Comment:
Considering that we're looking for the chain ID to change, I'm wondering whether we should wait for `NetworkController:networkDidChange` instead of `NetworkController:stateChange`. Otherwise, this will fire every time the NetworkController state changes (which is quite a lot).

Reply:
I was wondering how to integrate https://github.com/MetaMask/core/pull/3610! Fixed: https://github.com/MetaMask/core/pull/3609/commits/9c8491a0a98c512999b4f789b50c358c7e061df2